### ISSUE TITLE
Fix font aliasing

### ISSUE THREAD

**Issue by Otiel** [ForkIssues/TrackerWin#63]
I'm not sure if it's aliasing or something else but some fonts look choppy or clipped.

For instance, on this context menu, all characters of the same "value" are not rendered the same way and some are easier to read than others (look at the `r`, `l`, `F`...):

![image](https://user-images.githubusercontent.com/3851902/49162477-43b6d900-f32b-11e8-9eba-de0a962cffd9.png)

Or on this screenshot, look at the commit messages and paths:

![image](https://user-images.githubusercontent.com/3851902/49162588-782a9500-f32b-11e8-8fbf-1ec5c9ac14d8.png)

Are the fonts supporting ClearType?

**Comment by DanPristupov** [comment]
> Are the fonts supporting ClearType?

Yes, on 99% of systems (https://github.com/ForkIssues/TrackerWin/issues/12), but there are still some problems sometimes.

What OS, screen resolution, DPI?

**Comment by Otiel** [comment]
Windows 10 1809 (Build 17763.134), 1920x1080, 100%

**Comment by DanPristupov** [comment]
Do you have a single monitor or multiple ones?

**Comment by Otiel** [comment]
Two monitors, both 1920x1080.

**Comment by Otiel** [comment]
@DanPristupov Could you try to set `TextOptions.TextFormattingMode = "Display"` on small fonts (context menus, commits...) and see if the text looks clearer?

Have a look at [this resource](https://wpf.2000things.com/tag/textformattingmode/). He recommends the following:

>You should use the value of `Ideal`, except for the situations listed below.
>
>Set `TextFormattingMode` to `Display` when all of the following is true:
>
>* The FontSize of the text is 14 or less  (small text)
>* Text is not being transformed (scaled, rotated, translated)
>* The exact shape of the glyphs is not critical (e.g. for some graphic design scenario)

**Comment by Otiel** [comment]
@DanPristupov Did you had a chance to test this?

**Comment by vitalyster** [comment]
The problem is on 96dpi screens and `TextOptions.TextFormattingMode = "Display"` definitely fixes this

**Comment by DanPristupov** [comment]
I setup Windows 10 on an old laptop with 96 DPI  and going to investigate that.

**Comment by SylwesterZarebski** [comment]
Please fix this, all texts look awful because of greyscale AA.

**Comment by Otiel** [comment]
Tabs title even look blurry.

![image](https://user-images.githubusercontent.com/3851902/67688760-9f09a500-f99a-11e9-917f-d8ba87860914.png)

**Comment by SylwesterZarebski** [comment]
Is there RenderTransform used in any place? RenderTransform mixed with LayoutTransform (for zoom) breaks proper font rendering. Also if RenderTransform is used in any global Style it could be a cause of this.

It looks like interface at 100% is not really 100%, but scaled down from some bigger size.

Another thing is intermediate render targets (Effect, OpacityMask, VisualBrush, DrawingBrush, Clip, Opacity etc.), which also blocks ClearType rendering, but it looks like it is not a cause (RenderOptions.ClearTypeHint="Enabled" would force proper rendering on TextBlock, but i could not force to do that using WPFSnoop).

**Comment by Otiel** [comment]
@DanPristupov any chance to have this fixed soon? Some texts are really painful to read.

![image](https://user-images.githubusercontent.com/3851902/83237572-67afe080-a195-11ea-8ffa-bbd0e252eeab.png)

**Comment by vitalyster** [comment]
There is an another issue with both hi-dpi and 96-dpi screens - missing per monitor dpi awareness, texts became blurry when Fork dragged from hi-dpi screen to 96 dpi screen, when connected from 96dpi RDP session to Fork started on hi-dpi screen, etc.

**Comment by ma-bjo** [comment]
@DanPristupov Are you familiar with the SnapsToDevicePixels property that is available on most WPF controls? It usually works quite nice in order to get rid of the usual blurriness that WPF applications otherwise suffer from. We usually set it on the top level Form so that it affects the entire window.

Overriding the smoothing mode (ClearType etc.) during rendering is bad, applications should always respect the user's preference and go with the current mode configured in Windows.

**Comment by DanPristupov** [comment]
We use SnapsToDevicePixels of course.
The problem is that on all my computers (200 DPI, 96 DPI, TV) fonts look sharp.

**Comment by Otiel** [comment]
What about `TextOptions.TextFormattingMode`?

**Comment by vitalyster** [comment]
@DanPristupov what is the 96 dpi screen (model, vendor, scaling factor) you are using? Anyway, you will not loose anything if Display text formatting mode will be applied but it will help a lot of people :)

**Comment by ma-bjo** [comment]
The fonts look sharp on my machine as well, regular monitors with default dpi and no scaling. Could it be a problem with a corrupted WPF Font Cache on the machines that have this issue? That service has been known to cause blurry fonts on some machines, so maybe @Otiel can try clearing his cache?

**Comment by SylwesterZarebski** [comment]
[**@ma-bjo**]
> The fonts look sharp on my machine as well, regular monitors with default dpi and no scaling.

Hmm... Your images from issue #777 suffers from the same ugly greyscale AA. Maybe You are just used to this ugly AA behaviour, but it is very noticeable for me.

**Comment by Otiel** [comment]
>Could it be a problem with a corrupted WPF Font Cache on the machines that have this issue? That service has been known to cause blurry fonts on some machines, so maybe @Otiel can try clearing his cache?

This? https://winaero.com/blog/rebuild-font-cache-windows-10/

**Comment by ma-bjo** [comment]
Yes, that's the cache I was referring to. But as @SylwesterZarebski pointed out I do see what you meant also on my machine - the fonts are not using ClearType as they should. So please disregard my latest statement.

**Comment by vitalyster** [comment]
@ma-bjo 

> applications should always respect the user's preference and go with the current mode configured in Windows.

All applications which uses DirectWrite/Direct2D font rendering (as opposite to GDI rendering) are ignore ClearType "user preferences" by default and this is the problem described there.
With old GDI apps we can configure to use grayscale antialiasing (uncheck " enable ClearType") or subpixel antialiasing (check "enable ClearType") globally in Windows.
WPF uses DirectWrite and it uses grayscale AA by default. Subpixel AA in DirectWrite can only be enabled programmatically (D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE in C++, "TextFormattingMode=Display" in WPF).

**Comment by SylwesterZarebski** [comment]
@vitalyster : What? I've written few WPF apps and all of them uses cleartype/subpixel AA by default. TextFormattingMode=Display is not necessary for most situations. I'm using Ideal for all apps and mostly it is subpixel AA.

**Comment by vitalyster** [comment]
@SylwesterZarebski default WPF app without Display formatting mode looks exactly as Fork

**Comment by ma-bjo** [comment]
Whether fonts "look sharp" or "look blurry" can be very subjective. The easiest way to make sure is to take a screen dump and zoom in on the text using Photoshop or similar. If ClearType is used in the text rendering, there will be colored pixels also in grayscale text, but when (incorrect) font smoothing is used there is only grayscale pixels.

I just did that on Fork and cannot see that ClearType is used anywhere even though that is my preference in Windows.

**Comment by ma-bjo** [comment]
> @SylwesterZarebski default WPF app without Display formatting mode looks exactly as Fork

No, that is not true. I just tested by creating a new, empty WPF project for .NET Framework 4.8 and added a TextBlock to the default grid without changing anything. It renders in ClearType just as it should.

**Comment by vitalyster** [comment]
@ma-bjo 
> .NET Framework 4.8

Maybe that was changed recently, I will try it now. In this case we just need Fork built on 4.8 (and it also should fix per monitor dpi awareness too)

**Comment by DanPristupov** [comment]
We still use .Net 4.6.1 (.Net 4.6.2 will be used in the upcoming Fork 1.50).

**Comment by ma-bjo** [comment]
I rebuilt my test project for .NET 4.0 and it still uses ClearType as expected. Still **running** on 4.8 of course, but that is true also for Fork om my machine.

**Comment by vitalyster** [comment]
![text](https://user-images.githubusercontent.com/1052407/83262796-356aa700-a1c6-11ea-93c4-d0f71f8e290b.png)
Net 4.8 does not change anything, still need "Display" to see "crisp" text on 96DPI/100% scaling and still "blurry" without it (second block, Fork looks like that). Both blocks looks great on hidpi screen.

**Comment by ma-bjo** [comment]
@vitalyster Both text blocks in your example are using ClearType, so this is not the same issue as in Fork where ClearType is not being used at all. (Zoom in on the picture to see it.) But I agree with you that the first block is subjectively more readable.

**Comment by Otiel** [comment]
More tests:

![image](https://user-images.githubusercontent.com/3851902/83264881-0f8ed380-a1c1-11ea-9680-b83c32026529.png)

Zoomed in:

![image](https://user-images.githubusercontent.com/3851902/83265108-5977b980-a1c1-11ea-859e-a7b42d2413e7.png)

**Comment by SylwesterZarebski** [comment]
Setting TextFormattingMode=Display may bring back some contrast as seen above, but do not fix AA to be subpixel (although i have one or two cases when it works). Most greyscale AA is coming from transparency of grid or background. When background is solid and grid (or other) is not transparent, WPF should display subpixel/cleartype AA.

**Comment by SylwesterZarebski** [comment]
Also no subpixel AA on 1.50. Only menus inside have proper (system based) AA.

**Comment by DanPristupov** [comment]
We are investigating this issue.

@ma-bjo thank you very much for the the info. Now we can clearly differentiate blurry and sharp texts 👍.

@Otiel thank you for the tests.

It turned out that the problem is caused by a custom window chrome. It can be reproduced even in a simple project. Here's an example:

CustomWindow.xaml:
```
<local:CustomWindow x:Class="WpfApp5.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:local="clr-namespace:WpfApp5"
        Title="MainWindow" Height="450" Width="800">
    <StackPanel>
        <TextBlock FontSize="14" Text="Lock access to jobprofile events" Foreground="#222222"/>
    </StackPanel>
</local:CustomWindow>
```

CustomWindow.cs
```
public class CustomWindow : Window
{
    public CustomWindow()
    {
        WindowChrome.SetWindowChrome(this, new WindowChrome());
    }
}
```

I will share more info when we find how to fix this.

Update: looks like a bug in WPF: https://social.msdn.microsoft.com/Forums/vstudio/en-US/61242512-92c9-44ab-b59d-5f2c996d8ec9/windowchrome-and-textoptionstextformattingmodequotdisplayquot?forum=wpf

**Comment by SylwesterZarebski** [comment]
I wonder what Visual Studio is using to render its window. I'll try to find...

**Comment by vitalyster** [comment]
I have added the following:
```
UseLayoutRounding="True"
RenderOptions.BitmapScalingMode="NearestNeighbor"
SnapsToDevicePixels="True"
RenderOptions.ClearTypeHint="Enabled"
```
( + TextOptions.TextFormattingMode="Display")
And custom window now display text correctly

**Comment by vitalyster** [comment]
![def](https://user-images.githubusercontent.com/1052407/83524037-b2aa5a80-a4eb-11ea-976a-abcd357f871a.png) - default chrome
![customrounding](https://user-images.githubusercontent.com/1052407/83524050-b6d67800-a4eb-11ea-9dd1-e1a9ed9b48f7.png) - custom chrome
Looks absolutely identical after fixes

**Comment by DanPristupov** [comment]
Hi,

We made a build with many DPI-related changes. Could you try it and tell if you can see improvements?

The new version is available here as a zip archive (https://cdn.fork.dev/prerelease/Fork-1.50.5.zip). You need to extract it somewhere and run fork.exe.

**Comment by Otiel** [comment]
Looks better and easier on the eyes! 👀 

![image](https://user-images.githubusercontent.com/3851902/86025947-21b59900-ba2f-11ea-9da3-16894df45104.png)
![image](https://user-images.githubusercontent.com/3851902/86025855-03e83400-ba2f-11ea-93b1-b5655064f623.png)

It seems that your changes haven't been applied everywhere (yet?) though (for instance, on the File History window).

Thanks for working on this. 👍

**Comment by DanPristupov** [comment]
@Otiel thank you!

> It seems that your changes haven't been applied everywhere (yet?) though (for instance, on the File History window).

We will check the File History and Blame windows.

Does the diff control look properly?

**Comment by Otiel** [comment]
>Does the diff control look properly?

It's still grayscale AA:

![image](https://user-images.githubusercontent.com/3851902/86029596-a4d8ee00-ba33-11ea-9f75-eb5937f7ceb1.png)

**Comment by SylwesterZarebski** [comment]
Thanks Dan for great work! Not many people bring back proper subpixel AA in these days :-).

**Comment by DanPristupov** [comment]
Offtopic: it's not directly related, but recently there was a great post you guys may find interesting: https://tonsky.me/blog/monitors/.

**Comment by SylwesterZarebski** [comment]
Article is so much about Macs and their scaling and font rendering may be very good, but Windows has so inferior scaling:
- scaling is mostly unusable at other scales than integer ones (100%, 200% etc.), because system can't properly scale many programs (i.e. mmc.exe),
- even on these properly working programs, fonts are improperly scaled to some weird sizes (eg. font 9pt is not exact 13.5pt at 150%, but some other - closer to 14pt) and what is worse, they are still hinted to full pixels which makes them distorted; it is very visible to any person which also works with graphics and/or print.

Also:
- even 4K 27" monitor has clearly visible and better crispness with subpixel AA than greyscale AA,
- there are almost no available monitors with bigger resolutions than 4K (ignore few ones which costs tens of thousands of $).
- when someone also play games, he/she is tied to lower resolutions like WQHD or must buy very, very expensive GPU,

PS. 144Hz free/g-sync monitors gives more fun than higher resolution ;-).

**Comment by vitalyster** [comment]
1.50.5.0 is much better but still has issues: 

1) only some views now have improved font rendering (commits history, tab titles). Most of other views still have "blurry" fonts: branch tree, commit preview pane, etc. Button panel still have icons which are not snapped to device pixels and they are blurry as fonts too. All of these issues are only for 96 dpi screen, 100% scaling. On 192dpi and 200% all fonts and icons still looks crisp. 

2) Still no per monitor DPI awareness - when I drag Fork window from 96dpi monitor to 192dpi one - I have the same blurry fonts and icons even on hi-dpi screen!

**Comment by DanPristupov** [comment]
We created another build (1.50.6). https://cdn.fork.dev/prerelease/Fork-1.50.6.zip

There are many improvements everywhere (sidebar, diff text, windows, file history, blame etc). Could you try it tell if you see the difference and what is still blurry? Thank you!

> Still no per monitor DPI awareness - when I drag Fork window from 96dpi monitor to 192dpi one

We can't fix this in the _near_ future.

**Comment by ma-bjo** [comment]
Thanks Dan!

Now that more and more text is getting crisp I have started noticing that most icons seems to be half a pixel off as well, both in the sidebar to the left and the main toolbar at the top. Standard monitor, no high-dpi setting. It is mostly noticeable when there are straight lines in the graphics, such as for the folder icon for branches with / in them, and the icons for Changes, All Commits, Quick Launch and Stash.

This blurriness is also very prominent in lists and trees where file type, folder, and status icons are included, typically the lower parts of the Commit/Changes/File Tree tabs. For example, I imagine that the intention is that the ... and + Git status icons are squares with rounded corners, but the rounding is inconsistent and it looks irregular. They also have different sizes in the different tabs, so some of the blurring is most likely caused by scaling. There is room enough in all these lists to show a 16x16 px icon with no scaling.

I have also noticed that on the Commit tab, both the author and the commit is not properly aligned. The bold full name should have the base line aligned with the e-mail address that is shown immediately to the right of it. The e-mail needs to be moved 1px up for them to align properly. But now we are really taking details!

**Comment by vitalyster** [comment]
1.50.6 - much more improvements, but: context menu is blurry, commit pane is ok but its header with author and commit hash is blurry. All icons are blurry. Preferences window, quick launch window - blurry.
I tried to switch dark and light theme and noticed that these problems are lesser annoying in Dark theme, maybe most of users are using Dark theme and that is why they did not see problem :)

**Comment by SylwesterZarebski** [comment]
@vitalyster: Which context menu? I have subpixel AA in left bar and history context menus, also commit hash below has proper AA.
But, there is a but ;-). As @ma-bjo spotted, there is horizontal half-pixel off on everything and it is very visible on diff (i.e. letter m looks too bold).

What is not subpixel AA still:
- filter field in left bar
- keyboard shortcuts in menus
- line numbers
- filename/path in two-side diff
- in `Changes` tab view mode menu
- new branch window field
- new tag window fields
- commit subject and description (fields)
- Unstaged Changes text and `Stage` button in Changes tab
- Preferences window - almost all besides of diff preview and opened comboboxes entries

Button `Expand All`/`Collapse All` is partially over new changed commit files background.

**Comment by vitalyster** [comment]
@SylwesterZarebski interesting. Context menu are mostly ok except items with submenus. When item have submenu ("Interactive rebase ..." in commits, or "Tracking..." in branch tree) - it is "blurry"

**Comment by SylwesterZarebski** [comment]
It is subpixel AA in both submenus, but i think you just spot this horizontal half-pixel off what @ma-bjo spotted.
Probably this is main problem with blurriness, but with subpixel AA it is more noticeable on zoom - look at `f` or `i` letters or numbers (W10/4K/150%):
![obraz](https://user-images.githubusercontent.com/1755531/86367897-d1337b00-bc7c-11ea-8728-8c8a9fa33ca6.png)

Another one on another system (W10/WQHD/100%):
![obraz](https://user-images.githubusercontent.com/1755531/86386125-a9024700-bc91-11ea-92d4-07999d2f474d.png)

**Comment by SylwesterZarebski** [comment]
You can ignore above comment. It looks like Windows font rendering behaviour, very few applications has better font rendering and it looks like all of them has own font rendering (i.e. Firefox with natural/natural symmetric setting).
Some application has better contrast on fonts, but i do not know if this is possible in WPF.

**Comment by SylwesterZarebski** [comment]
Mostly Fork has very nice AA, only few places could be changed like history window - left list is mostly greyscale AA.

**Comment by DanPristupov** [comment]
@SylwesterZarebski can you make a screenshot, please? What monitor do you use, what resolution and DPI?

Does anyone else see that issue?

**Comment by SylwesterZarebski** [comment]
It looks like this:
![obraz](https://user-images.githubusercontent.com/1755531/228967902-d190b5a1-3367-42e0-908d-f81fff35da63.png)
I've made color adjustment to better show area of greyscale AA on the left.

PS. It is very minor issue, thank You for great work :).

**Comment by DanPristupov** [comment]
@SylwesterZarebski we just released 1.83 update. Please check for updates.

**Comment by netinput** [comment]
Is there a way to completely disable antialiasing? TextFormattingMode.Display surely does the trick, would be super great to have it in the settings.

**Comment by SylwesterZarebski** [comment]
> @SylwesterZarebski we just released 1.83 update. Please check for updates.

Thanks a lot, now it is fine.
Last greyscale AA i've found is in update window, but it is very minor:
![obraz](https://github.com/fork-dev/TrackerWin/assets/1755531/5b7a5c11-cdbc-471e-8787-70acca247bd1)

**Comment by ma-bjo** [comment]
> Is there a way to completely disable antialiasing? TextFormattingMode.Display surely does the trick, would be super great to have it in the settings.

Applications should not have their own configuration for antialiasing, they should just follow the configuration that the user has done in Windows.

**Comment by netinput** [comment]
> Applications should not have their own configuration for antialiasing, they should just follow the configuration that the user has done in Windows.

I have it completely disabled on my system, because it gives me some eye strain. Especially now when the greyscale is replaced with rainbowish fonts. But WPF doesn't follow the system settings. I've seen [other apps](https://gist.github.com/tamr/301a4a3ad23a5dd39f359dfc1cae3a56) having this issue, but there's no solution for closed source.

**Comment by SylwesterZarebski** [comment]
`TextFormattingMode.Display` do not disable AA nor obey Windows settings, just set text formatter to use *GDI-compatible font metrics*, which can make things better or worse depending on your setup.
As @babanga has written above, WPF doesn't follow system settings and there is little you can do.
One way is to buy very high DPI screen, which minimise AA effects (also subpixel).

**Comment by WesselKroos** [comment]
Hi @DanPristupov, thanks for all the work you've done so far on improving the readability of text!

However, in 2 places the readability is worse for me with ClearType turned on in Windows 11. Which are the line numbers and the description of hidden lines:

![image](https://github.com/fork-dev/TrackerWin/assets/31220528/b0d496e5-3b62-49c5-adeb-4d9351dae6b0)

When I compare the AA effect with a similar text in the Google Chrome browser it looks as if the effect is rotated by 180 degrees and over-saturated:

| | Fork | Google Chrome |
|--|--|--|
| **Windows ClearType off** | ✅| AA is still on? |
| **Windows ClearType on** | rotated by 180 degrees & over-saturated? | ✅ |

![image](https://github.com/fork-dev/TrackerWin/assets/31220528/8818d02c-bec6-47d8-a668-887c10453a45)

Could these be reasons why these texts look more blurry in comparison with ClearType turned off?

**Comment by SylwesterZarebski** [comment]
@WesselKroos : It looks like reversed RGB, thus wrong cleartype. Have You tried dark theme?

**Comment by WesselKroos** [comment]
@SylwesterZarebski I think you are onto something, looking at dark- vs light mode.

The selected ClearType is the first one in this selection screen, which is sharper on my monitor. I guess that is RGB? because the second ClearType displays a blue line at the left, and a faint yellow line at the right side of the T (the first character):

![image](https://github.com/fork-dev/TrackerWin/assets/31220528/be4ade8a-2a34-434c-8fe3-421e1c2de6c0)

But dark mode looks way sharper:

![image](https://github.com/fork-dev/TrackerWin/assets/31220528/0551fa78-0514-44d7-abd6-3bcbd8f0a076)

Notice how the colors in the filepath `src/script/libs/amb...` has a clear reversed/rotated AA effect while the linenumbers, description of hidden lines and file contents stay almost the same but should be reversed/rotated:

![image](https://github.com/fork-dev/TrackerWin/assets/31220528/faa82dca-05ba-4cd8-abff-3272db508b8e)

It looks like the text within this scrollable window has a wrong/reversed/rotated ClearType in light mode:

![image](https://github.com/fork-dev/TrackerWin/assets/31220528/b1d55202-18ba-4160-bdb1-2b71a0d4e709)

**Comment by SylwesterZarebski** [comment]
Very weird. Only Dan could shed some light on this case.
